### PR TITLE
MM-14196 Add virtualised list for posts

### PR DESCRIPTION
**Pull request by sudheerDev**
#### Summary

**Known issues**

- [x] P1: First team switch does not reliably work because of channel sync issue.
- [x] slow network fast posting causes scroll pops upwards
- [x] text box expansion should not overlap post list content
- [ ] compact view last post jumping (due to hover menu)

**QA testing cases**

Performance with lots of data:
- busy channel with many incoming posts - verify no white space when scrolling or sitting at the bottom of the channel
- Scroll near the top of a channel with 5000+ messages and then try to switch channels. Examine perf
- Post a message when scrolled near the top of a channel with 5000+ messages
- Click "new messages below" indicator at the top of a channel with lots of posts.

Worst case scenarios:
- Editing posts and adding reactions to most recent post and posts up in the channel. Verify no flickering or scroll pop
- Channels with images, message attchements, webhook message attachments, markdown, link previews and youtube previews, long posts. Verify no pop during post, scrolling or refresh

Scrolling smoothness:
- Scroll quickly in one direction then quickly switch scrolling direction.
- Permalinks to posts in channels that aren't loaded yet
- test various browsers
- various network speeds

Misc:
- Permalinks
- Jump to message from search
- channel and team switching. Compare to community

____

Completed:

- [x] Link preview and SVG can throw off scroll correction a little. That will be addressed with https://github.com/mattermost/mattermost-webapp/pull/2334 and https://github.com/mattermost/mattermost-webapp/pull/2469
- [x] youtube previews cause LARGE pop
- [x] removing posts can cause blank space at the top of the channel
- [x] channel header missing on channels with few posts
- [x] In progress: Fix issue of flicker when a a new reaction is added (Does not happen always most likely related to the way style cache is busted).
- [x] scroll channel down to bottom when current user make a new post
- [x] Permalink view should move post to middle of screen and highlight
- [x] Remove padding below channel header
- [x] logic tweak for bottom of the channel so we never see unmounted posts (now shows 10 max)
- [x] Fix jumping to permalink when posts are not loaded yet
- [x] Scroll to bottom when adding reaction or editing most recent post
- [x] Youtube previews causing minor incorrect scroll correction
- [x] Don't scroll center channel to bottom when posting in the RHS
- [x] Loader can get stuck at the top of the screen if we can repro
- [x] minor scroll bar adjustment when entering the channel on subsequent visits
- [x] increase number of posts to 100 (in scrolling direction)- 50 (in opposite direction)
- [x] increase load threshold to 1000px
- [x] consecutive posts are being separated in center channel
- [x] stop loading two pages on first load (only load 30 posts)
- [x] deleting system messages leaves blank space
- [x] fix overlapping issue when making every alternate posts in a channel that has a lot of posts loaded
- [x] Manual loadtest of posts to verify this issue os only related to the /test command: Fix loading indicator stuck and showing blank posts above (see image below)

____

#### Ticket Link
[MM-14196](https://mattermost.atlassian.net/browse/MM-14196)
#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Touches critical sections of the codebase (auth, posting, etc.)

### CODE REVIEW

**Comment by esethna** [comment]
Thanks Sudheer, this is looking really good. Some things I noticed:

1. White bar at the top:

image

2. Some channels are infinitely stuck in the loading more posts state (try in "autem" and "doloremque"): 
image

Edit: actually all channels seem to be doing this at some point far enough back in the channel

3. The above ^ makes it a bit hard to test but I feel like we are loading much too early. Notice how the scroll bar starts to shrink almost immediately as I scroll up:

scroll

As far as I know the rules we put in place last time involved loading more posts at X% from the top. If you test Slack it has a much nicer feel because they seem to only load when you are almost at the top of the loaded posts (however I would say that they are loading a bit later than I would like us to do).

4. I did not notice any unmounted posts (white space). Is the virtualized post list active or is it just done well so I'm not noticing it?

**Comment by esethna** [comment]
Added new to do list items to PR description ^

**Comment by esethna** [comment]
Updated PR description with new P1 to dos before QA testing

**Comment by esethna** [comment]
Updated the PR description with new fixes required

**Comment by DHaussermann** [comment]
@sudheerDev When I open a channel with a collapsed image attachment and expand it, the position breaks. Seems to only happen near the top of a channel though. https://i-008032a64c672fb74.test.spinmint.com/ad-1/pl/fozizd4yz7yh8pkgebmqonhgxo

**Comment by ogi-m** [comment]
@sudheerDev Chrome - Expanding a youtube link on the bottom of the page positions the preview out of view, the view isn't scrolled to the bottom (not an issue on Firefox where the view is scrolls to the bottom)
expand_yt

**Comment by ogi-m** [comment]
@sudheerDev Firefox - glitch when expanding a youtube link that's not on the bottom of the channel. 
Note in the gif how the preview shows up on top of the url before taking it's correct position below the url:
expand_yt1

**Comment by ogi-m** [comment]
@sudheerDev Clicking on a [permalink](https://i-008032a64c672fb74.test.spinmint.com/ad-1/pl/rx3wcej697r13b76id8tapu3to) opens a white screen.

**Comment by esethna** [comment]
Discussed with QA. Before merging to master we want to fix high priority issues, then work on the other known issues as it soaks on community-daily.

P1 (before merging to master - needs new spinmint):
- Fix regression with jumpiness on loading new posts
- Fix permalink view + jumping to post from search

P2 (after merging to master):
- Investigate lag on safari
- Small delay on clicking New Messages below bubble (1 second)
- https://github.com/mattermost/mattermost-webapp/pull/2447#issuecomment-475445716
- https://github.com/mattermost/mattermost-webapp/pull/2447#issuecomment-475456037
- https://github.com/mattermost/mattermost-webapp/pull/2447#issuecomment-475462621

cc// @sudheerDev @DHaussermann @ogi-m

File: components/create_post/index.js [hunk 1]
```diff
@@ -28,7 +28,6 @@ import {Posts, Preferences as PreferencesRedux} from 'mattermost-redux/constants
 
 import {connectionErrorCount} from 'selectors/views/system';
 
-import {postListScrollChangeToBottom} from 'actions/global_actions.jsx';
```
[the remark this answers is not in the document]
Comment (sudheerDev) [reply components/create_post/index.js]: Yup i was waiting to check if i have to remove globalevent emitter as well along with this. Removed both now.

File: components/post_view/index.js [hunk 2]
```diff
@@ -3,21 +3,51 @@
...
 
 import {increasePostVisibility} from 'actions/post_actions.jsx';
 import {checkAndSetMobileView} from 'actions/views/channel';
-import {Preferences} from 'utils/constants.jsx';
+import {makePreparePostIdsForPostList} from 'selectors/posts';
+import {Constants} from 'utils/constants.jsx';
 
 import PostList from './post_list.jsx';
 
+// This function is added add a fail safe for the channel sync issue we have.
```
[the remark this answers is not in the document]
Comment (sudheerDev) [reply components/post_view/index.js]: done.

File: components/post_view/message_attachments/message_attachment/message_attachment.jsx [hunk 3]
```diff
@@ -4,8 +4,6 @@
 import PropTypes from 'prop-types';
 import React from 'react';
 
-import {postListScrollChange} from 'actions/global_actions';
```
[the remark this answers is not in the document]
Comment (sudheerDev) [reply components/post_view/message_attachments/message_attachment/message_attachment.jsx]: Yup remove these.

**Comment by sudheerDev** [comment]
@saturninoabril @enahum completed functional changes. This is up for review. I am adding some testes for post_list.jsx but other than that feel feel to review.

File: components/post_view/index.js [hunk 4]
```diff
@@ -3,21 +3,48 @@
...
+            }
+        }
+        if (channel.team_id && channel.team_id !== team.id) {
+            return true;
+        }
+    } else {
+        return true;
+    }
+
+    return false;
```
Comment (enahum) [inline components/post_view/index.js]: non blocking but perhaps this `return false` can be placed inside the if statement and the return true outside of the else? like this

```js
const isChannelLoading = (...) => {
    ...
    if (channel && team} {
       ...
       return false;
    }

    return true;
}

**Comment by sudheerDev** [comment]
@esethna @DHaussermann @ogi-m 
P1 (before merging to master - needs new spinmint):
- Fix regression with jumpiness on loading new posts - Done
- Fix permalink view + jumping to post from search - Done

P2 (after merging to master):
- Investigate lag on safari - Partially done but we can improve this 
- Small delay on clicking New Messages below bubble (1 second)
- https://github.com/mattermost/mattermost-webapp/pull/2447#issuecomment-475445716
- https://github.com/mattermost/mattermost-webapp/pull/2447#issuecomment-475456037 - Done
- https://github.com/mattermost/mattermost-webapp/pull/2447#issuecomment-475462621 - Done

File: components/post_view/index.js [hunk 5]
```diff
@@ -3,21 +3,48 @@
...
+};
```
Comment (saturninoabril) [inline components/post_view/index.js]: Non-blocking but the last part is a bit confusing to read (at least for me).  Could it be simplified?  Any chance that it's same as,
```javascript
const isChannelLoading = (params, channel, team) => {
    if (params.postid) {
        return false;
    }

    if (channel && team) {
        if (channel.type !== Constants.DM_CHANNEL && channel.type !== Constants.GM_CHANNEL) {
            if (channel.name !== params.identifier) {
                return true;
            }
        }
        if (channel.team_id && channel.team_id !== team.id) {
            return true;
        }

        return false;
    }

    return true;
};
```

File: components/post_view/post/post.jsx [hunk 6]
```diff
@@ -64,9 +64,9 @@ export default class Post extends React.PureComponent {
         replyCount: PropTypes.number,
 
         /**
-         * Function to get the post list HTML element
+         * The post count used for selenium tests
          */
-        getPostList: PropTypes.func.isRequired,
+        lastPostCount: PropTypes.number,
```
Comment (saturninoabril) [inline components/post_view/post/post.jsx]: This `lastPostCount` can be safely removed as it was removed from master.

File: components/post_view/post_list.jsx [hunk 7]
```diff
@@ -42,6 +30,18 @@ export default class PostList extends React.PureComponent {
          */
         posts: PropTypes.array,
 
+        /**
+         * Array of Ids in the channel inlcuding date separators, new message indicator, more messages loader,
```
Comment (saturninoabril) [inline components/post_view/post_list.jsx]:
```suggestion
         * Array of Ids in the channel including date separators, new message indicator, more messages loader,
```

File: components/post_view/post_list.jsx [hunk 8]
```diff
@@ -440,194 +260,168 @@ export default class PostList extends React.PureComponent {
...
+    itemKey = (index) => {
+        const {postListIds} = this.state;
+        return postListIds[index] ? postListIds[index] : index;
+    }
 
-            this.updateFloatingTimestamp();
+    onScroll = ({scrollDirection, scrollOffset, scrollUpdateWasRequested}) => {
+        const isNotLoadingPosts = !this.state.postsLoading && !this.loadingPosts;
+        const didUserScrollBackwards = scrollDirection === 'backward' && !scrollUpdateWasRequested;
+        const isOffsetWithInRange = scrollOffset < 1000 || window.scrollOfsetHelper; //initial offset will be 0 before the desired first scroll position
```
Comment (saturninoabril) [inline components/post_view/post_list.jsx]: Can't see reference to `window.scrollOfsetHelper`.  Also, there's a typo.

File: components/post_view/index.js (hunk 4 above)
Comment (saturninoabril) [reply components/post_view/index.js]: lol, I had this same comment.

File: components/post_view/index.js (hunk 5 above)
Comment (sudheerDev) [reply components/post_view/index.js]: Yup, Makes sense. Fixed it

File: components/post_view/post_list.jsx (hunk 8 above)
Comment (sudheerDev) [reply components/post_view/post_list.jsx]: Nice catch 👍 . I forgot about this

File: components/post_view/index.js (hunk 4 above)
Comment (sudheerDev) [reply components/post_view/index.js]: Changed it. That was a result of iterating when i was too tired :)

**Comment by esethna** [comment]
@amyblais removing the docs needed label. In my mind infinite scroll/virtualized post list is expected functionality from a users perspective, rather than something that needs to be called out in docs. Feel free to re-add if you think there's something we should cover specifically.

**Pull request merged**